core-common/src/main/java/eu/arrowhead/common/processor/NetworkAddressDetector.java

```diff
@@ -0,0 +1,212 @@
+	public AddressDetectionResult detect(final HttpServletRequest servletRequest) { // TODO junit
+		logger.debug("detect started...");
+		final AddressDetectionResult result = new AddressDetectionResult();
+		
+		if (!useDetector || servletRequest == null) {
+			result.setSkipped(true);
+			result.setDetectionMessage("Address-Port detection process was skipped");
+			return result;
+		}
+		boolean detected = retrieveFromConnector(servletRequest, result);
+		
+		if (!detected) {
+			detected = detectByHeaderForwarded(servletRequest, result);			
+		}
+		if (!detected) {
+			detected = detectByHeaderXForwardedFor(servletRequest, result);
+		}
+		
+		if (detected) {
+			try {
+				networkAddressVerifier.verify(result.getDetectedAddress()); // already normalized with NetworkAddressPreProcessor
+				return result;
+			} catch (final InvalidParameterException ex) {
+				result.setDetectionSuccess(false);
+				result.setDetectionMessage("Detected address is invalid! " + ex.getMessage());
+				return result;
+			}
+		} else {
+			result.setDetectionMessage("Network address detection was unsuccessful.");
+			return result;
+		}		
+	}
+	
+	//-------------------------------------------------------------------------------------------------
+	@PostConstruct
+	private void init() {
+		for (final String addr : filterProxyAddressList) {
+			filterProxyAddressSet.add(networkAddressPreProcessor.normalize(addr));
+		}
+	}
+	
+	//-------------------------------------------------------------------------------------------------
+	private boolean retrieveFromConnector(final HttpServletRequest servletRequest, final AddressDetectionResult result) {
+		String remoteAddr = networkAddressPreProcessor.normalize(servletRequest.getRemoteAddr());
+		
+		if (remoteAddr.equals(NetworkAddressVerifier.IPV4_PLACEHOLDER) ||
+			remoteAddr.startsWith(NetworkAddressVerifier.IPV4_LOOPBACK_1ST_OCTET) ||
+			remoteAddr.equals(NetworkAddressVerifier.IPV6_LOOPBACK) ||
+			remoteAddr.equals(NetworkAddressVerifier.IPV6_UNSPECIFIED)) {
+			
+			remoteAddr = networkAddressPreProcessor.normalize(servletRequest.getLocalAddr());
+		}
+		
+		if (Utilities.isEmpty(remoteAddr) || filterProxyAddressSet.contains(remoteAddr)) {
+			return false;
+		} else {
+			result.setDetectedAddress(remoteAddr);
+			result.setDetectionSuccess(true);
+			return true;
+		}		
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private boolean detectByHeaderForwarded(final HttpServletRequest servletRequest, final AddressDetectionResult result) {
+		//https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded
+		//https://tools.ietf.org/html/rfc7230#section-3.2
+		final List<String> headerValues = Collections.list(servletRequest.getHeaders(HEADER_FORWARDED));
+		for (int i = headerValues.size() - 1; i >= 0; i--) {
+			
+			String[] subValues = new String[1];
+			if (headerValues.get(i).contains(SEMI_COLON)) {
```

Comment:
can't be both comma and semi-colon?

Reply:
According to the examples listed [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded) and the RFC within I have assumed that no.

Reply:
ok, I didn't know.

Reply:
I found this in the internet

> If a HttpRequest with a header of Forwarded: for=192.168.0.55;proto=https;host=example.com,for=172.19.0.1;proto=http is passed to ProxyHeaderParser
> 
> Expected Behaviour
> 
> Based on the spec, it should break it apart by comma first, into two parts, one for each proxy (for=192.168.0.55;proto=https;host=example.com and for=172.19.0.1;proto=http), and then for each part, break it apart by a semi-colon. 

https://github.com/micronaut-projects/micronaut-core/issues/4571

Reply:
Seems you are right, this also confirms: https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/